쥬스 메이커 [STEP 1] mmim, marisol

JoSH0318 opened the pull request:
@stevenkim18 안녕하세요 ☺️ 첫 PR 입니다 ! 잘부탁드립니다.
---
## UML [STEP 1]
<img width="3328" alt="UML" src="https://user-images.githubusercontent.com/88810018/155549280-00a142e9-b476-463a-89c8-d400d09657bd.png">

## 고민한점 / 해결방법
1. 어떻게 한 메서드만을 이용하여 선택된 다른 Juice마다 `fruitInventory`에 변화를 줄지에 대한 고민을 정말 오랜시간 했습니다.
   - Juice가 요구하는 과일, 수량이 recipe에서 [Fruit: Int]로 선언하였습니다. (필요로 하는 과일 종류가 2가지인 경우가 있어서ㅜㅜ) Juice를 선택하면 선택된 Juice에 따라 recipe가 필요로하는 과일들과 수량을 제공하고 자동으로 `fruitInventory`을 차감시키기 위해 `makeJuice`메서드를 구현했습니다. 
2. Dictionary타입의 타입 프로퍼티를 선언했는데, 이때 타입 프로퍼티의 Initialization에 대한 고민을 했습니다.
   - 재고의 초기값을 위한 `static let defaultQuantity = 10` 선언하고, `fruitInventory`의 Value 값에 기입했습니다. 이게 좋은 방법은 아닌것 같습니다. 아쉬움이 많이 남는 코드입니다. 조언 부탁드립니다!!!
   ```swift
    static let defaultQuantity = 10
    static var fruitInventory: Dictionary = [Fruit.strawberry: defaultQuantity,
                                             Fruit.banana: defaultQuantity,
                                             Fruit.pineapple: defaultQuantity,
                                             Fruit.kiwi: defaultQuantity,
                                             Fruit.mango: defaultQuantity]
   ```
3. JuiceMaker 구조체의 makeJuice() 메서드의 for문 내부에서 value를 사용하지 않는데도 인수를 (key, value) 튜플로 받는 이유가 무엇일지 궁금합니다.
```swift
    func makeJuice(by recipe: [Fruit: Int]) throws {
        for (fruit, _ ) in FruitStore.fruitInventory {
            guard let fruitStock = FruitStore.fruitInventory[fruit] else {
                throw ErrorHandling.wrongFormat
            }
            guard let requiredQuantity = recipe[fruit] else {
                continue
            }
            guard fruitStock >= requiredQuantity else { throw ErrorHandling.outOfStock }
            FruitStore.fruitInventory[fruit] = fruitStock - requiredQuantity
        }
    }
```
4. 에러 핸들링하는 메서드의 경우 에러가 맞느냐, 정상 동작하느냐로 나뉜다고 생각합니다. 그런 측면에서 true/false를 나타내는 bool타입 네이밍과 비슷하다고 생각이 들었습니다. 이런 경우 bool타입 네이밍과 같은 방법으로 네이밍을 해도 될지 궁금합니다.
5. 에러 핸들링을 위한 enum 타입의 이름을 ErrorHandling으로 지었는데, 너무 범용적으로 사용되는 이름이라는 생각을 했습니다. 고민 끝에 결론이 나지 않았는데 조언 부탁드립니다.

JoSH0318 committed:
[docs]: README.md update

- 팀 그라운드 룰 초안 업데이트 : 스크럼, 프로젝트, 커밋, 브랜치 규칙

marisol-develop committed:
[feat]: add enum FruitType

JoSH0318 committed:
[feat]: Add fruitList: Dictionary, func updateQuantity()

- fruitList: 초안으로 Dic에 모든 키값의 value를 10으로 지정
- func updateQuantity(): view controller에서 받아온 값을 파라미터로 받고
   해당 과일 키값의 수를 업데이트

marisol-develop committed:
[feat]: add class Recipe, func selectMenu() in JuiceMaker

- class Recipe에 주스별로 메서드 생성
- selectMenu()에서 switch문 돌며 주스별로 레시피 메서드 실행

JoSH0318 committed:
[feat]: makeStrawberryJuice() 기능 구현, Recipe.swift 삭제

marisol-develop committed:
[feat]: add enum ErrorHandling

에러 처리해주는 열거형 ErrorHandling 생성

JoSH0318 committed:
[feat]: add makeJuice() method

- 음료의 레시피 정보를 받아 음료별로 재고 변경, 에러 처리 해주는 makeJuice()메서드 추가

marisol-develop committed:
[feat]: enum ErrorHandling 생성, checkOrder() 메서드 생성

- 오류처리하는 ErrorHandling 열거형 생성
- checkOrder()메서드에서 do-catch로 오류 처리
- 네이밍 수정

JoSH0318 committed:
[docs]: add STEP1 UML in README.md

marisol-develop committed:
[docs]: STEP1 고민한점, 배운개념 작성

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/ErrorHandling.swift:
```diff
@@ -0,0 +1,24 @@
+
+import Foundation
+
+enum ErrorHandling: Error {
+    case outOfStock
+    case wrongFormat
+}
```
Error를 사용하셔서 enum을 만들어 보셨네요!! 훌룡합니다!!!

JoSH0318 replied:
👍

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/ErrorHandling.swift:
```diff
@@ -0,0 +1,24 @@
+    case wrongFormat
+}
+
+extension ErrorHandling: LocalizedError {
+    var description: String {
+        switch self {
+        case .outOfStock:
+            return "재고가 부족합니다."
+        case .wrongFormat:
+            return "잘못된 입력입니다."
+        }
+    }
+}
```
`description` 를 정의하시고 사용을 하지 않으신 것 같은데 언제 사용될까요??

JoSH0318 replied:
사용하고 있습니다! do-catch 문에서 catch의 statement에서 사용하고 있습니다.
이 후에 알럿을 통한 메시지로 활용할 것 같습니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Fruit.swift:
```diff
@@ -0,0 +1,31 @@
+
+import Foundation
+
+enum Fruit {
+    case strawberry
+    case banana
+    case pineapple
+    case kiwi
+    case mango
+    
+    var name: String {
+        switch self {
+        case .strawberry:
+            return "딸기"
+        case .banana:
+            return "바나나"
+        case .pineapple:
+            return "파인애플"
+        case .kiwi:
+            return "키위"
+        case .mango:
+            return "망고"
+        }
+    }
+}
```
Enum을 사용하신 것도 너무 잘하셨어요!! name이라는 연산 프로퍼티를 정의하신 이유는 무엇인가요??

JoSH0318 replied:
이 후에 다음 스텝에서 사용하려고 만들었습니다. 하지만 이번 스텝의 요구사항에 없기에 빼도록 하겠습니다 😂
ex) "딸기"의 재고가 부족합니다.

stevenkim18 replied:
`name` 이라는 연산 프로터티를 정의하실 수도 있지만 rawValue를 사용해서 값을 나타낼 수도 있을 것 같아요
enum에 rawValue도 한번 공부해보세요!

JoSH0318 replied:
rawValue를 사용하는 간단한 방법이 있지만, 후에 언급할 때 `Fruit.banana.rawValue`라고 언급하게 되어 rawValue가 이름인지, 수량인지, 또는 그외의 어떠한 값인지 추측이 되지 않는 단점이 있다고 판단했습니다. 지금과 같은 경우는 같은 값이 없지만 같은 값을 rawValue를 사용할 수 없는 단점도 있는 것 같습니다. 
만약 지금의 코드에서 언급하게 될 때 `Fruit.banana.name`으로 접근하게 되어, 이름을 언급하는 구나 추축이 가능하다고 생각합니다.

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,11 @@
 
-// 과일 저장소 타입
 class FruitStore {
-    
+    static let defaultQuantity = 10
+    static var fruitInventory: Dictionary = [Fruit.strawberry: defaultQuantity,
+                                             Fruit.banana: defaultQuantity,
+                                             Fruit.pineapple: defaultQuantity,
+                                             Fruit.kiwi: defaultQuantity,
+                                             Fruit.mango: defaultQuantity]
 }
```
이부분은 꼭 수정해주시고
왜 class를 사용했는지
static은 언제 사용하는지를 공부해보세요!!!

marisol-develop replied:
FruitStore의 Inventory가 참조타입이어야 하기 때문에 class를 사용해야 한다고 생각합니다.
static var는 특정 유형의 모든 인스턴스에 보편적인 값을 정의할 때 유용하다고 배웠습니다!
많은 인스턴스를 생성하더라도 모든 인스턴스에 고정된 값이 필요할 때 써야 하는데, 저희의 경우 값이 업데이트 되어야 하는데 static var를 써버렸더라구요 🥲

이 부분은  inventory 변수에 static을 제거하고, 이니셜라이저를 통해 초기화 하도록 변경했습니다.
```swift
var inventory: [Fruit: Int] = [:]
    init() {
        let defaultQuantity = 10
        Fruit.allCases.forEach{ inventory[$0] = defaultQuantity }
    }
```

stevenkim18 replied:
잘 수정하셨습니다!! 계속 많이 사용해보시면서 클래스와 구조체의 쓰임새와 특징을 숙지해주세요!!! 기본중에 기본입니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,11 @@
 class FruitStore {
-    
+    static let defaultQuantity = 10
+    static var fruitInventory: Dictionary = [Fruit.strawberry: defaultQuantity,
+                                             Fruit.banana: defaultQuantity,
+                                             Fruit.pineapple: defaultQuantity,
+                                             Fruit.kiwi: defaultQuantity,
+                                             Fruit.mango: defaultQuantity]
```
앞에 `Fruit` 생략 가능합니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/Juice.swift:
```diff
@@ -0,0 +1,37 @@
+    case mangoJuice
+    case mangoKiwiJuice
+    
+    var recipe: [Fruit: Int] {
+        switch self {
+        case .strawberryJuice:
+            return [.strawberry: 16]
+        case .bananaJuice:
+            return [.banana: 2]
+        case .kiwiJuice:
+            return [.kiwi: 3]
+        case .pineappleJuice:
+            return [.pineapple: 2]
+        case .strawberryBananaJuice:
+            return [.strawberry: 10, .banana: 1]
+        case .mangoJuice:
+            return [.mango: 3]
+        case .mangoKiwiJuice:
+            return [.mango: 2, .kiwi: 1]
+        }
+    }
```
튜플도 있는데 딕셔너리를 사용하신 이유가 궁금합니다!! 어떤 장단점이 있을까요??

JoSH0318 replied:
1. 튜플은 Multiple Return Values를 위해 사용하는 것으로 생각합니다. 
2. Dictionary를 사용하는 경우 key값만으로 value에도 접근 가능해지는 장점이 있습니다.
3. 딸기바나나쥬스와 같이 두가지 과일의 각각의 수량 정보를 담기 위해서는 배열이 필수적이라고 생각했습니다. 배열중에서 한 쌍으로 데이터를 관리할 수 있는 Dictionary가 적합하다고 생각합니다.
4. value 값에 접근할 때, Dictionary를 사용하면 `recipe[.strawberry]`로 접근할 수있어서 가독성이 좋습니다. 하지만 튜플을 사용할 시 `recipe[0].quantity`과 같이 접근해야 해서 어떤과일의 수량에 접근하는지 알 수 없습니다.(tuple name을 사용했다고 가정)

이러한 근거로 Dictionary를 사용했습니다 😅 혹시 저희가 잘못 생각한 부분이 있을까요?

stevenkim18 replied:
정답은 없다고 생각합니다!! 튜플과 딕셔너리 각각의 장단점이 있습니다!!!
사용을 했을 때 그냥 사용하신 것이 아닌 이유가 있으면 좋습니다!! 너무 엇나가는 이유라면 제가 지적해드릴게요!!
잘 고민해보신 것 같아요!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,48 @@
 import Foundation
 
-// 쥬스 메이커 타입
 struct JuiceMaker {
```
여기는 왜 구조체로 되어있을까요??

marisol-develop replied:
우선 `JuiceMaker`에서 값이 참조타입이어야 하는 프로퍼티가 없습니다.
`JuiceMaker`는 따로 상속을 받거나, 상속을 해줄 필요가 없다고 생각했습니다.
이러한 경우 stack에 값을 할당하는 struct가 속도가 빠르다고 배웠기 때문에 struct로 되어 있다고 생각합니다.

혹시 struct로 되어 있는 중요한 이유를 놓친거라면, 말씀 부탁드립니다.🙇‍♀️

stevenkim18 replied:
제 개인적인 생각에는 주스 메이커는 클래스를 사용해도 큰 문제는 없을 것 같아요!

클래스에 `final` 키워드를 붙이면 상속을 받지 못하게 할 수 있습니다!

구조체를 사용한 이유가 단순히 속도 때문이라면 
클래스와 구조체의 속도차이가 얼마나 나고 
지금 같이 메모리가 많이 제공되는 시점에서 유의미한 차이가 있을까요??

marisol-develop replied:
우선 제가 리뷰어님께서 여쭤보신 질문에 정확한 답을 할 정도의 수준이 못되는 것 같습니다.
정확히 클래스와 구조체의 속도 차이가 얼마나 나는지는 모르겠습니다. 다만 구조체 타입을 변수에 할당하면, Copy-On-Write 방법으로 일단은 메모리를 할당하지 않고 같은 곳을 보고, 해당 값을 변경할 때 그때서야 실제로 메모리에 값을 복사하고 값을 변경하는 걸로 알고 있습니다. 이를 통해 메모리 최적화가 가능하고, 클래스에 비해 성능이 좋다고 하는 것 같습니다. 다만 저희의 경우 해당 구조체 내부에 Class 프로퍼티를 포함하고 있기 때문에 메모리 최적화가 어려울 것 같긴 하네요.🥲
메모리가 넉넉히 제공되는 시점에서 이러한 고민이 유의미한 차이가 없을지도 모르겠습니다. (더구나 작은 프로젝트니까요) 하지만 작은 프로젝트를 할 때부터 이런 고민을 하면 좋지 않을까 하는 생각이 들었습니다.

저번 주에 swift의 Structures and Classes라는 문서를 읽었습니다. Choose Structures by Default라고 되어 있더라구요. “클래스가 지원하는 추가 기능으로 인해 복잡성이 증가한다. 일반적으로는 구조체를 선호하고, 필요한 경우 클래스를 사용하는 것을 권장한다.“고 나와있고, 클래스의 “상속”이라는 기능 또한 프로토콜을 통해 구현하여 채택해서 사용하는 방법을 제안하더라구요.
그래서 final을 굳이 써가면서 class를 사용해야 하는 이유를 찾지 못했습니다. 추후에 class로 변경해야하는 상황이 온다면 그때 고려해보도록 하겠습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,48 @@
 
-// 쥬스 메이커 타입
 struct JuiceMaker {
+    var fruitStore = FruitStore()
```
이 프로퍼티는 외부에서 접근 해야되나요??

위와 같이 초기화하는 것과 이니셜라이져에서 초기화하는 것은 어떤 차이가 있을까요??

JoSH0318 replied:
1. 외부에서 접근할 일이 없어보입니다! `private` 키워드로 접근제어하도록 수정했습니다.
2. 위와 같이 초기화하면 `class FruitStore`의 `init`에서 설정해놓은 default 값으로 초기화됩니다. 만약 이니셜라이져에서 초기화하면 매번 해당값을 설정해줘야합니다. 이번 JuiceMaker와 같은 경우는 변하지 않는 10이라는 default 값이 존재하기 때문에 전자가 더 효율적이라고 생각했습니다!

stevenkim18 replied:
1. 보통은 프로퍼티를 접근 제어하는 것이 좋지만 설계에 따라 달라질 수 있기 떄문에 꼭 private를 사용하실 필요는 없습니다.
2. 잘 찾으셨습니다!! 지금 시점에서는 조금 이르지만 나중에 각각 언제 메모리에 생성 되는지도 공부해보시면 좋을 것 같습니다!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,48 @@
 struct JuiceMaker {
+    var fruitStore = FruitStore()
     
+    func selectMenu(_ juice: Juice) {
+        switch juice {
+        case .strawberryJuice:
+            checkOder(Juice.strawberryJuice.recipe)
+        case .bananaJuice:
+            checkOder(Juice.bananaJuice.recipe)
+        case .kiwiJuice:
+            checkOder(Juice.kiwiJuice.recipe)
+        case .pineappleJuice:
+            checkOder(Juice.pineappleJuice.recipe)
+        case .strawberryBananaJuice:
+            checkOder(Juice.strawberryBananaJuice.recipe)
+        case .mangoJuice:
+            checkOder(Juice.mangoJuice.recipe)
+        case .mangoKiwiJuice:
+            checkOder(Juice.mangoKiwiJuice.recipe)
+        }
+    }
```
이 함수는 많이 짧게 작성할 수 있을 것 같은데 어떻게 하면 될까요??

JoSH0318 replied:
앗! 신기하게 코드 짤 때는 안보이던게 이제야 보이는 느낌입니다 🥲
case별로 바뀌는 로직이 없고 받는 인자만 바뀌니 단 한줄로 표현되네요 😭
수정했습니다.

stevenkim18 replied:
pr을 보내실 때는 구현이 완료 되었다고 해서 바로 보내시지 마시고 
다시 한번 수정할 것은 없나 리펙토링할 것읍 없나 검토해보세요!!

기능을 구현하는 것 보다 검토하는 것이 휠씬 중요합니다!!

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,48 @@
+        }
+    }
+    
+    func checkOder(_ recipe: [Fruit: Int]) {
+        do {
+            try makeJuice(by: recipe)
+        } catch let error as ErrorHandling {
+            print(error.description)
+        } catch {
+            print(error)
+        }
+    }
```
do try catch문을 사용하신 것은 정말 잘하셨어요!!

try와 try? 의 차이는 무엇인가요??

JoSH0318 replied:
제 생각에는 try는 컴파일러에게 "이 메서드가 정상 동작할 수도, 에러를 낼 수도 있어"라고 말해주는 키워드라고 생각합니다.
따라서 에러를 낼 수 있는 여지가 있는 메서드 앞에 붙여주고, 
정상 작동할 때는 do, 에러일 때 catch를 통해 제어한다고 생각합니다!

try? 같은 경우에는 메서드에 에러가 나면 `nil`값을 기본적으로 반환해줍니다. 따라서 do-catch문 없이도 사용됩니다.
☺️

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,48 @@
+        }
+    }
+    
+    func checkOder(_ recipe: [Fruit: Int]) {
```
오타가 있어요!!!

JoSH0318 replied:
수정했습니다 ㅎㅎ 😅

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,48 @@
+    }
+    
+    func makeJuice(by recipe: [Fruit: Int]) throws {
+        for (fruit, _ ) in FruitStore.fruitInventory {
+            guard let fruitStock = FruitStore.fruitInventory[fruit] else {
+                throw ErrorHandling.wrongFormat
+            }
+            guard let requiredQuantity = recipe[fruit] else {
+                continue
+            }
+            guard fruitStock >= requiredQuantity else { throw ErrorHandling.outOfStock }
+            FruitStore.fruitInventory[fruit] = fruitStock - requiredQuantity
+        }
```
throw 하는 부분을 44번째 줄에서는 개행하셨고 49번쨰 줄에서는 그렇지 않은 이유가 궁금합니다!!

JoSH0318 replied:
수정했습니다 ㅎㅎ 😅

stevenkim18 replied:
코드에는 일관성이 있어야 합니다!!! 다음 pr에서는 꼭 신경써주세요!

stevenkim18 reviewed (commented):
안녕하세요! 

밈 마리솔 오히려 제가 더 잘 부탁드립니다!! 그리고 늦어져서 죄송합니다 ㅜㅜ 다음주까지 조금 바쁠 것 같지만 그래도 최대한 빨리 해드리려고 노력할게요!! 사파리 레드 팀이 먼저 리뷰를 주셔서 그곳 먼저 하고 오느라 늦었습니다!

질문 많이 해주시면 감사할게요!! 저도 찾으면서 공부하는 시간을 가질 수 있어요!

일단 리뷰에 들어가기 전에 말씀 드리고 싶은게 있어요!!

저의 말은 정답이 아닙니다! 코딩에는 정답이 없습니다! 

그래서 제가 어떤 질문을 하거나 리뷰를 남기면 무조건적으로 수용하지 마시고

- 항상 “왜" 라는 질문을 많이 하시면서 많이 고민해주세요!
- 제가 만약 틀렸다면 근거를 가지고 반박해 주세요!
- 혹 더 좋은 방법이 있다면 제시해주세요!

처음에는 어려우시겠지만 더 많이 생각하시고 고민하시는 개발자가 되시기를 바래요!

한쪽으로만 생각해보셨다면 다른 방법으로도 접근해 보시고 다른 관점에서도 생각 해보시면 좋을 것 같습니다!

--- 

가장 먼저 여쭤보고 싶은 것은 **이 코드가 동작을 하는지가 궁금합니다**. 실제로 주스메이커를 만들어서 함수를 호출하면 재고가 증감이 되는지가 궁금합니다. 이 부분을 다시 구현하셔서 커밋해주시면 감사하겠습니다!!

⚠️⚠️⚠️⚠️⚠️⚠️⚠️⚠️

또한 두분에게 물어보고 싶은 것은... 다음과 같은 개념들을 잘 이해하고 계신지를 여쭤보고 싶습니다.

- Class와 Struct의 차이
- static은 언제 사용하는지
- 지역 변수와 전역 변수의 차이
- OOP가 무엇인지
- 타입, 인스턴스, 프로퍼티가 무엇인지

위의 개념들을 정확히 이해하고 저한테 설명해 주시면 좋을 것 같아요!!

```swift
class FruitStore {

     static let defaultQuantity = 10
     static var fruitInventory: Dictionary = [Fruit.strawberry: defaultQuantity,
                                              Fruit.banana: defaultQuantity,
                                              Fruit.pineapple: defaultQuantity,
                                              Fruit.kiwi: defaultQuantity,
                                              Fruit.mango: defaultQuantity]
 }
```

위와 같은 개념을 공부해보시면 `FruitStore` 을 다시 만들어 보실 수 있으실 것 같아요!!! 전혀 감이 잡히지 않으시면 또 질문해주세요!!

추가적으로 리드미에 보면 uml을 그리신것 같은데 pr을 보낼 때 같이 첨부해주시면 더 좋을 것 같아요!! 어떻게 하면 우리가 작성한 코드를 잘 보여줄 수 있을지도 고민해보시면 좋을 것 같습니다!

## PR에 대한 답변

### 2. Dictionary타입의 타입 프로퍼티를 선언했는데, 이때 타입 프로퍼티의 Initialization에 대한 고민을 했습니다.

- 이 부분은 제가 위해서 요구 드린 것을 잘 공부해보시고 나서 고민해 보시는 것이 좋을 것 같습니다.

### 3. JuiceMaker 구조체의 makeJuice() 메서드의 for문 내부에서 value를 사용하지 않는데도 인수를 (key, value) 튜플로 받는 이유가 무엇일지 궁금합니다.

- 두분께서 정의 하신 `fruitInventory` 가 딕셔너리라서 그렇습니다!!

### 4. 에러 핸들링하는 메서드의 경우 에러가 �맞느냐, 정상 동작하느냐로 나뉜다고 생각합니다. 그런 측면에서 true/false를 나타내는 bool타입 네이밍과 비슷하다고 생각이 들었습니다. 이런 경우 bool타입 네이밍과 같은 방법으로 네이밍을 �해도 될지 궁금합니다.

- 질문이 정확히 무엇을 물어보시는 건지 잘 모르겠어요 ㅜㅜ
- `makeJuice` 를 말씀하시는 건가요?
- 아니면 `ErrorHandling` 타입 자체를 말씀하시는건가요?

### 5. 에러 핸들링을 위한 enum 타입의 이름을 ErrorHandling으로 지었는데, 너무 범용적으로 사용되는 이름이라는 생각을 했습니다. 고민 끝에 결론이 나지 않았는데 조언 부탁드립니다.

- `ErrorHandling` 은 진행형인 느낌이 있어서 저러면 CustomError 같이 Error가 뒤에 들어가는 이름으로 지을 것 같아요!
- 이름 짓는 부분 관련해서는 많이 검색해보시고 [[Swift API 디자인 가이드에 네이밍 파트](https://www.swift.org/documentation/api-design-guidelines/#naming)](https://www.swift.org/documentation/api-design-guidelines/#naming)를 꼭 읽어보세요

JoSH0318 committed:
[fix]: delete static keyword and add initializer

delete static keyword and add initializer으로 인한 추가적인 코드 수정

JoSH0318 committed:
Merge branch '5_mmim90' of github.com:JoSH0318/ios-juice-maker into 5_mmim90

marisol-develop committed:
[refactor]: 네이밍 수정, selectMenu() 수정

- checkOrder() -> order()로 수정
- ErrorHandling -> OrderError로 수정
- selectMenu() 코드 수정

JoSH0318 committed:
[improve]: modify logic in makeJuice() method

- 비효율적인 로직 개선

marisol-develop committed:
[feat]: add changeFruitStock() in FruitStore, enum ButtonType

- 과일의 수량을 n개 변경하는 changeFruitStock() 생성
- +, - 를 구별하는 ButtonType 열거형 생성

JoSH0318 committed:
[rename]: rename property, method, parameter, enum, enum case

marisol-develop commented:
@stevenkim18
안녕하세요 스티븐! 피드백 주신 부분 반영하여 수정했습니다.🙇‍♀️
확인 부탁드립니다!☺️

## 추가 구현 내용
- `func updateStock()`: do-catch문으로 구성되어있는 메서드 / 과일재고의 increase, decrease를 판단하고 오류를 다루는 역할
- `func increaseStock()`: 과일 재고를 증가시키는 메서드
- `func decreaseStock()`: 과일 재고를 감소시키는 메서드
- `enum Button {}`: stepper button의 plus, minus를 구분

--------------------------------------

## 고민한 점
- FruitStore 클래스에서 과일의 수량을 변경하는 메서드 `updateStock(by inputQuantity: [Fruit: Int], button: Button)`을 구현했습니다. button이 plus일 때는 `increaseStock(by inputQuantity: [Fruit: Int])`를 실행하고, button이 minus일 때 `decreaseStock(by inputQuantity: [Fruit: Int])`를 실행하도록 코드를 작성했습니다. increaseStock()과 decreaseStock() 두 메서드 모두 알 수 없는 에러를 검사하는 로직이 동일하게 적용되어서, 이 두 메서드를 하나로 합칠 수 없을지 고민했습니다. for문 안에 if문을 쓰면 들여쓰기가 두 번이 되어서, 우선 두 메서드를 분리해서 작성해두었습니다.


--------------------------------------
### Class와 Struct의 차이
- class는 참조 타입이고, struct는 값 타입입니다. class는 Heap에 할당되고, struct는 stack에 할당됩니다.
- 클래스의 경우, 같은 클래스의 인스턴스를 여러 개의 변수에 할당한 뒤 값을 변경시키면 모든 변수가 변경되고, 구조체는 같은 구조체의 인스턴스를 여러 변수에 할당 한 뒤 값을 변경시키더라도 다른 변수에 영향을 주지 않습니다.
- class는 상속이 가능하고, struct는 상속이 불가합니다.
- struct는 멤버와이즈 이니셜라이저 기능이 제공되지만, class는 상속 기능 때문에 멤버와이즈 이니셜라이저 기능이 제공되지 않습니다.
- 타입 캐스팅은 class의 인스턴스에서만 사용 가능합니다.
- deinitializer도 class에서만 사용 가능합니다.

###  static은 언제 사용하는지
- 특정 유형의 모든 인스턴스에 보편적인 값을 정의하는데에 유용합니다. 그렇기 때문에 많은 인스턴스를 생성하더라도, 모든 인스턴스에 고정된 값이 필요할 때 사용합니다.

###  지역변수와 전역변수의 차이
- 지역변수는 함수나 메서드 블럭 내에서 선언되며, 선언된 함수나 메서드 외부에서는 사용할 수 없습니다. 지역변수가 선언된 블럭이 종료될 때 메모리에서 삭제됩니다.
- 전역변수는 모든 함수의 외부에 선언되며, 모든 함수 및 메서드를 통해 액세스 할 수 있습니다. 프로그램이 종료될 때 메모리에서 삭제됩니다. 

###  OOP가 무엇인지
- OOP는 프로그램을 어떻게 설계해야 하는지에 대한 방법론 중 하나이며, 객체 지향 프로그래밍을 말합니다. (Object-Oriented Programming) 프로그래밍을 명령어의 목록으로 보는 시각에서 벗어나, 여러 개의 독립된 단위 (객체)들의 상호작용으로 프로그램의 동작을 구현합니다.
- 세상의 모든 것을 객체로 표현하며, 객체들의 유기적인 관계를 통해 프로세스가 진행됩니다.
- 추상화, 캡슐화, 상속, 다형성의 특징을 갖습니다.

###  타입, 인스턴스, 프로퍼티가 무엇인지
- 타입은 설계도이고, 인스턴스는 설계도인 타입에 따라 만든 실체입니다.
- 프로퍼티는 클래스 또는 구조체의 인스턴스와 연관된 값을 저장 또는 연산합니다.

marisol-develop committed:
[refactor]: JuicaMaker 내부의 fruitStore 인스턴스 private 처리

JoSH0318 committed:
[docs]: modify UML [STEP1] in README.md

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -6,7 +6,46 @@
+        }
+    }
+    
+    func increaseStock(by inputQuantity: [Fruit: Int]) throws {
+        for (fruit, quantity) in inputQuantity {
+            guard let fruitStock = inventory[fruit] else {
+                throw OrderError.unknownError
+            }
+            inventory[fruit] = fruitStock + quantity
+        }
+    }
+    
+    func decreaseStock(by inputQuantity: [Fruit: Int]) throws {
```
이 함수들이 외부에서 사용되나요?

stevenkim18 commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -6,7 +6,32 @@
+        }
+    }
+
+    func makeJuice(by recipe: [Fruit: Int]) throws {
```
이 함수도 마찬가지로 외부에서 사용되나요?

stevenkim18 reviewed (approved):
수정해주신 코드 잘 봤습니다!!

고민하신 흔적이 보여서 좋았습니다!!

다만 전체적으로 보면서 조금 더 드리고 싶은 말이 있습니다.

### 대부분의 예외처리를 do-try-catch만으로 구현하셨어요!
왜 모든 예외처리를  do-try-catch만으로 처리하셨나요? 다른 방법은 없을까요?
있다면 바른 방법은 없을까요??

error 케이스가 두가지 밖에 없는데 do-try-catch를 많이 사용하셔서 오히려 코드가 더 보기 어려워진 느낌입나다.

### indent 2번이상 코드가 너무 많아요!
프로젝트 귄고사항에 들여쓰기는 최대한 함수안에서 한번만 하라고 되어있습니다.
`updateStock` 내부에서는 들여쓰기가 3번이 되어있습니다. 
이부분을 신경써서 코드를 작성해보시면 많은 도움이 되실 것 같아요!!

### 고민하신 점에 대한 답변
일단 함수를 나눠서 구현하신 것은 잘하셨습니다.
만약 하나의 함수로 합칠 수 있다면 
저라면 증감여부를 알려주는 bool 타입의 매개변수를 추가할 것 같습니다.

### 기본 문법을 설명해주신 것에 대한 답변
구두로 말했듯이 단순히 개념만을 안다고 다 아는 것 아닙니다.
이번 과제에서 끊임 없이 뒤 주제를 가지고 많이 고민하시고 공부해주시면 감사하겠습니다!!

--- 

머지를 해드리겠습니다!! 
제가 댓글 달아 드린 것에 대해서 추가적으로 고민해 주시고 
다음 pr은 많이 공부해보시고 고민해보시고 주시면 감사하겠습니다!!
(목표가 목욜까지로 되어있으니 너무 빨리 안주셔도 됩니다!! 
더 많이 고민해보고 다른 캠퍼들이랑 의견을 많이 나눠보는 것이 중요합니다!!)

중간에 질문이 있으시면 언제든지 DM 주세요!
